### PR TITLE
Subscriptions implementation

### PR DESCRIPTION
**Pull request by RigidStudios**
Awaiting review from @Inctus and @RigidStudios

### CODE REVIEW

File: src/lib/tree/nodes/Subscription.ts [hunk 1]
```diff
@@ -1,49 +1,70 @@
...
 	FAILED,
 }
 
-export type KeyedHandler<T> = [subtype, T];
+export class NodeSubscription<T> {
+	protected handlers: Array<[subState, SubscriptionCallback<T, unknown>]> = new Array();
+	private returnsStack: never[] = [];
+
+	private errMsg = "";
+	private failed = false;
```
Comment (RigidStudios) [inline src/lib/tree/nodes/Subscription.ts]: Why does this exist as a class field?

File: src/lib/tree/nodes/Subscription.ts [hunk 2]
```diff
@@ -1,49 +1,70 @@
...
+							const lastReturn = this.returnsStack[0];
+							const arg = lastReturn !== undefined ? lastReturn : v;
 
-	/** @hidden */
-	public fire() {}
+							this.returnsStack.pop();
+							this.returnsStack.push(handler(arg) as never);
+						} catch (e) {
+							this.errMsg = e as string;
+							this.failed = true;
```
Comment (RigidStudios) [inline src/lib/tree/nodes/Subscription.ts]: ^

File: src/lib/tree/nodes/Subscription.ts [hunk 3]
```diff
@@ -1,49 +1,70 @@
...
+enum subState {
 	THEN,
 	CATCH,
 	FAILED,
 }
 
-export type KeyedHandler<T> = [subtype, T];
+export class NodeSubscription<T> {
+	protected handlers: Array<[subState, SubscriptionCallback<T, unknown>]> = new Array();
+	private returnsStack: never[] = [];
```
Comment (RigidStudios) [inline src/lib/tree/nodes/Subscription.ts]: Again, why the field?

File: src/lib/tree/nodes/Subscription.ts [hunk 4]
```diff
@@ -1,49 +1,70 @@
...
+
+	private errMsg = "";
+	private failed = false;
+
+	public fire(v: T): void {
+		const handlers = this.handlers;
```
Comment (RigidStudios) [inline src/lib/tree/nodes/Subscription.ts]: Don't alias

File: src/lib/tree/nodes/Subscription.ts [hunk 5]
```diff
@@ -1,49 +1,70 @@
...
+				case subState.THEN:
+					if (!this.failed) {
+						try {
+							const lastReturn = this.returnsStack[0];
+							const arg = lastReturn !== undefined ? lastReturn : v;
 
-	/** @hidden */
-	public fire() {}
+							this.returnsStack.pop();
+							this.returnsStack.push(handler(arg) as never);
```
Comment (RigidStudios) [inline src/lib/tree/nodes/Subscription.ts]: Why exactly is there even a stack? Why not just keep a fixed storage location?

**Comment by RigidStudios** [comment]
@siriuslatte View the above ^

**Comment by RigidStudios** [comment]
I might suggest just not doing subscriptions entirely and letting people pass functions? @siriuslatte thoughts?

**Comment by RigidStudios** [comment]
It's not a bad structure by any means to make subscriptions but not everyone will want to structure them like this, so who's to say we should force them instead of allowing `(val) => { /* do stuff */ }`

**Comment by haashim-ali** [comment]
Have decided to implement subs more simply, allowing end users to add whatever abstractions they want on top.

**Pull request closed without merging**